x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSort.java

```diff
@@ -0,0 +1,243 @@
+            );
+            case INT -> new MvSort.EvaluatorFactory(
+                toEvaluator.apply(field),
+                ordering,
+                (blockFactory, fieldBlock, sortOrder) -> new MultivalueDedupeInt((IntBlock) fieldBlock).sortToBlock(blockFactory, sortOrder)
+            );
+            case LONG -> new MvSort.EvaluatorFactory(
+                toEvaluator.apply(field),
+                ordering,
+                (blockFactory, fieldBlock, sortOrder) -> new MultivalueDedupeLong((LongBlock) fieldBlock).sortToBlock(
+                    blockFactory,
+                    sortOrder
+                )
+            );
+            case DOUBLE -> new MvSort.EvaluatorFactory(
+                toEvaluator.apply(field),
+                ordering,
+                (blockFactory, fieldBlock, sortOrder) -> new MultivalueDedupeDouble((DoubleBlock) fieldBlock).sortToBlock(
+                    blockFactory,
+                    sortOrder
+                )
+            );
+            case NULL -> EvalOperator.CONSTANT_NULL_FACTORY;
+            default -> throw new IllegalArgumentException("unsupported type [" + field.dataType() + "]");
+        };
+    }
+
+    private record EvaluatorFactory(
+        EvalOperator.ExpressionEvaluator.Factory field,
+        boolean order,
+        TriFunction<BlockFactory, Block, Boolean, Block> sort
+    ) implements EvalOperator.ExpressionEvaluator.Factory {
+        @Override
+        public EvalOperator.ExpressionEvaluator get(DriverContext context) {
+            return new MvSort.Evaluator(context.blockFactory(), field.get(context), order, sort);
+        }
+
+        @Override
+        public String toString() {
+            return "MvSort[field=" + field + ", order=" + order + "]";
+        }
+    }
+
+    private static class Evaluator implements EvalOperator.ExpressionEvaluator {
+        private final BlockFactory blockFactory;
+        private final EvalOperator.ExpressionEvaluator field;
+        private final boolean order;
+        private final TriFunction<BlockFactory, Block, Boolean, Block> sort;
+
+        protected Evaluator(
+            BlockFactory blockFactory,
+            EvalOperator.ExpressionEvaluator field,
+            boolean order,
+            TriFunction<BlockFactory, Block, Boolean, Block> sort
+        ) {
+            this.blockFactory = blockFactory;
+            this.field = field;
+            this.order = order;
+            this.sort = sort;
+        }
+
+        @Override
+        public Block eval(Page page) {
+            try (Block fieldBlock = field.eval(page)) {
+                return sort.apply(blockFactory, fieldBlock, order);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "MvSort[field=" + field + ", order=" + order + "]";
+        }
+
+        @Override
+        public void close() {}
+    }
+
+    @Override
+    public Object fold() {
+        return EvaluatorMapper.super.fold();
```

Comment:
If you extend `EsqlScalarFunction` then you get this for free. I think that's possible here and saves a tiny bit of thinking so I think it's worth it.